### DISCUSSION TITLE
Working with nested Outputs

### DISCUSSION THREAD

**Discussion by crabhi** [pulumi/pulumi#12696] [Pulumi in Practice] [answered]
Hi everyone,
after I create a stack, I need to generate a custom config file. For this, I need to access some outputs, such as `my_list_of_ec2_instances[i].public_ip`.

I haven't found a good way to wait for all nested outputs to be available and then access them. Especially because I have a nested structure of objects. So far, the best way seems to use `Output.json_dumps`:

```python
my_config = { 
   ...   # using lifted outputs from the generated objects
}
Output.json_dumps(my_config).apply(lambda x: json.loads(x)).apply(something_something)
```

Is there a way to do the above without the JSON trip?

**Comment by Frassle** [comment, marked as answer]
The typing information is sub-optimal for this operation but if you just call `Output.from_input(my_config)` it will collapse output values to plain values recursively.

**Reply by crabhi** [reply]
Works like a charm, thanks!

I didn't get it from the docs that this function can transform a (potentially nested) `Dict[str, Output[T]]` into `Output[Dict[str, T]]`.

**Reply by crabhi** [reply]
By the way, why is `json_dumps` needed then? Couldn't it be implemented as `Output.from_input(data).apply(json.dumps)`?

**Reply by Frassle** [reply]
`from_input` only recurses through known types (tuples, lists, dictionaries, etc) so if you have a custom type with Output attributes on it `from_input` won't change those. It could be surprising if you had a class explicitly with an Output value on it which would get made into a plain value if it got passed through `from_input`.
`json_dumps` will handle nested Output values even in custom types, this is safe because the final result is just a JSON string.

**Reply by crabhi** [reply]
Interesting, thank you!
